roxie/roxiemem/roxiemem.cpp

```diff
@@ -5207,18 +5257,24 @@ void DataBufferBottom::addToFreeChain(DataBuffer * buffer)
 
 void DataBufferBottom::Release()
 {
-    if (atomic_dec_and_test(&count))
+    if (count.fetch_sub(1, std::memory_order_release) == 1)
+    {
+        //No acquire fence - released() is assumed not to access the data in this buffer
+        //If it does it should contain an acquire fence
         released();
+    }
 }
 
 void DataBufferBottom::released()
 {
     // Not safe to free here as owner may be in the middle of allocating from it
     // instead, give owner a hint that it's worth thinking about freeing this page next time it is safe
-    if (atomic_cas(&count, DEAD_PSEUDO_COUNT, 0))
+    unsigned expected = 0;
+    if (count.compare_exchange_strong(expected, DEAD_PSEUDO_COUNT, std::memory_order_release))
     {
-        atomic_set(&owner->freePending, 1);
-        atomic_set(&okToFree, 1);
+        //No acquire fence required since the following code doesn't read anything from the object
+        okToFree.store(true, std::memory_order_release);
```

Comment:
I made this, and the compare_exchange release - because the isAlive() test might have failed otherwise.